Basic Implementation of Plugins and Plugin Functions

jhwgh1968 opened the pull request:
At long last, here it is: plugins as I originally envisioned them.

In order to achieve this, I had to also add "lazy modules" as a concept. That is, modules who defer lookups on their contents until they are accessed. Currently such modules can only be implemented in plugins,  but that is an accident of the implementation. I tried to make sure the concept could be extended later if it was useful.

The way I integrated "lazy modules" into the existing module code is rather ugly. Suggestions or cleanup fixes afterwords are welcome (so long as my one new test does not break).

schungx commented:
I've made some comments into the code.  The No.1 suggest I can make is not to separate the two module types, but to unify them, if it is possible at all.  For example, make normal modules a subset of a super `Module` type - they just eager-load instead of lazy load.  The user, and the rest of the program, doesn't really need to care.  This way, it provides maximum flexibility to the future.

If the two module types cannot be merged (I don't see why we can't), then at least merge them into one single `Union` type.  Then you avoid most of the duplicated code in the rest of the program.  For example:

```
enum SuperModule {
    EagerModule(Module),
    LazyModule(LazyLoadedModule)
}

Union::Module(SuperModule)
```

This way, the code can be simplified as:

```
let module = scope.get_module(index)........

match module {
    Dynamic(Union::Module(EagerModule(m))) => { ... }
    Dynamic(Union::Module(LazyModule(m))) => { ... }
}
```

schungx commented:
Please look at https://github.com/schungx/rhai/pull/13

That PR adds an abstraction `NativeCallable` which you can then implement for your lazy-loaded functions to hook directly into the function calling mechanism.

jhwgh1968 commented:
I finally have time to get back to this.

I have just rebased onto #13, and will make the two changes you suggest soon.

jhwgh1968 commented:
> The No.1 suggestion I can make is not to separate the two module types, but to unify them, if it is possible at all.

I agree, but I created the implementation separately, because the current API provided by modules is too complex to merge.

1. The hashes are returned to other parts of the engine, as part of the public API of the `Module` struct.
2. There is not only `set_fn`, but then `set_fn_1` and `set_fn_1_mut`.
3. There are a ton of utility functions like `contains_fn` and `index_all_modules`, which rely on everything being dynamic.

I would like to bridge the two implementations by making `Module` into a trait, but that seems tricky. I can only imagine something like:

```
enum LookupResult {
    StaticallyLoaded(u64), // statically loaded, the hash used for lookups
    DynamicallyLoaded // return code meaning: call lazily instead
}

trait Module {
    // the lookup functions
    fn lookup_var(name: &str) -> Option<&Dynamic>;
    fn lookup_fn(name: &str, args: Iterator) -> Option<LookupResult>;
    fn lookup_recv_fn(name: &str, first_arg: &mut Dynamic, args: Iterator) -> Option<LookupResult>;

    // the dynamic invocation functions. non-lazy modules may return None
    fn call_fn(name: &str, args: Iterator) -> Option<Result<Dynamic, EvalAltResult>>;
    fn call_recv_fn(name: &str, first_arg: &mut Dynamic, args: Iterator) -> Option<Result<Dynamic, EvalAltResult>>;

    // return a table of static items. dynamic modules return an empty one
    fn get_static_items(&self) -> &HashMap<u64, Box<dyn NativeCallable>>;
}
```

Achieving that is a lot more rework than I think I could achieve. I have not dug far enough into the way modules work, except to see that it's really complex and code across multiple places.

What do you think? Do you have extra groundwork you can do to help me? Or should I take another approach to unifying them?

jhwgh1968 commented:
Oh, and `NativeCallable` doesn't help me, because that is just a return type for the lookup functions. I need a trait for the actual looking up that wraps/abstracts _the entire lookup process, including the hashing_.

jhwgh1968 commented:
I did read your other thread, where you said:

>> Aside from being simpler for me to understand, having a code-based lookup -- call this function, don't worry about what it does -- allows better flexibility and options for optimization in the future.
> This is a great idea. We can simply modify FnAny to be a struct that implements that trait and have all module functions return a trait object. [...] The more I think about this, the more this is beautiful. 

So I'm presuming that this means you don't want hashing to be part of the public API for module-like items, even if (as you say there), lazy-loaded modules still do that. Or is this incorrect?

If it is, then I worry I don't have enough bandwidth to get much further on this without more guidance...

schungx commented:
That's right.  I consider hashing an internal implementation details to speed up lookup's of loaded functions.  It is an internal indexing step.

The plugin writer really should not care about the hashes.  They should provide the normal information to the engine: function name, list of parameter types, and the ABI type.  The engine should load the plugin file, index all the functions internally using this information.

So I'd expect your plugins API to expose those data instead of directly hashing them.  The engine can hash.

Or, alternatively, I can provide a standard black-box API function to do the hashing and return a key for the plugin function.  But I still think it is best to keep the hashing internal to the engine, as we can then have plugins compiled for different versions of Rhai.

jhwgh1968 commented:
That makes sense. However, the current modules implementation is difficult for me to understand.

If I understand correctly, you are thinking a trait like this could be used by the engine:

```
trait Module {
    /// Iterate over the functions inside this module known at load time.
    ///
    /// Some modules may be lazily loaded, so not all functions in a module may be returned.
    fn static_fns<'s>(&'s self) -> Box<Iterator<Item=NativeCallable> + 's>;

    /// Iterate over the constants inside this module known at load time.
    ///
    /// Some modules may be lazily loaded, so not all constants in a module may be returned.
    fn static_consts<'s>(&'s self) -> Box<Iterator<Item=ScopeEntry> + 's>;

    /// Look up a function dynamically.
    ///
    /// Functions may pass a first argument as a mutable receiver. If this is not done, the second argument is None.
    ///
    /// May return a callable which was not in `static_fns`, if the function was lazily loaded.
    fn lookup_fn(&self, receiver_arg: Option<&mut Dynamic>, args: &[Dynamic]) -> Option<NativeCallable>;

    /// Look up a constant dynamically.
    ///
    /// May return a callable which was not in `static_fns`, if the function was lazily loaded.
    fn lookup_fn(&self, mut_first_arg: Option<Dynamic>, args: &[Dynamic]) -> Option<NativeCallable>;
}
```

With this, the engine could still do the hashing in advance for static modules, and if there was no hash found, try a dynamic call before deciding, "module not found."

If that sounds good, it would really help me if you could make current modules in `src/modules.rs` work by implementing this trait, or something similar. I could then do the unification after a rebase.

schungx commented:
Yeah, something like that.  However, the actual function names in the trait probably should match what already exists so the code doesn't need to change much, only changing from `Module` type to a trait object of `dyn Module`.

Also, I have removed the `NativeCallable` trait based on your observations.  Reintroduce it if you think it is necessary, or otherwise just return a `NativeFunction` type instead, which is a simple wrapper of a Boxed function pointer and an ABI type enum.

I always have difficulty understanding the 'lazy-loading" bit and why it prevents things from being pre-generated and pre-compiled.  Rust is a compiled language, so when the "lazy-load" module is loaded, it already contains all the function implementations.  There really is nothing to "lazy-load", except for the module library file itself.

I think I consistently fail to understand why handles cannot simply be pre-generated (or generated once during module loading time) for _all_ functions in the module.  You just need to avoid having your proc macro generate a large `match` statement on function names and argument types, but instead generate code to hash the function, then insert it into a `HashMap`.  A simple implementation is to use a `Vec` to hold all the macro-generated functions.  The "handle" in this case is simply the `Vec` offset.   When I load your module for the first time, pass me a reference to that `Vec` and I can index everything in one go.

schungx commented:
> 
> 
> Yeah, something like that. However, the actual function names in the trait probably should match what already exists so the code doesn't need to change much, only changing from `Module` type to a trait object of `dyn Module`.
> 
> Also, I have removed the `NativeCallable` trait based on your observations. Reintroduce it if you think it is necessary, or otherwise just return a `NativeFunction` type instead, which is a simple wrapper of a Boxed function pointer and an ABI type enum.
> 
> I always have difficulty understanding the 'lazy-loading" bit and why it prevents things from being pre-generated and pre-compiled. Rust is a compiled language, so when the "lazy-load" module is loaded, it already contains all the function implementations. There really is nothing to "lazy-load", except for the module library file itself.
> 
> I think I consistently fail to understand why handles cannot simply be pre-generated (or generated once during module loading time) for _all_ functions in the module. You just need to avoid having your proc macro generate a large `match` statement on function names and argument types, but instead generate code to hash the function, then insert it into a `HashMap`. A simple implementation is to use a `Vec` to hold all the macro-generated functions. The "handle" in this case is simply the `Vec` offset. When I load your module for the first time, pass me a reference to that `Vec` and I can index everything in one go.   Something like:

```rust
trait Module {
    :
   fn get_functions(&self) -> Box<dyn Iterator<Item = FnSpec>>;
    :
}
```

schungx commented:
So, a restructure will first introduce a `Module` trait.

Then the existing `Module` type will be split into separate types:

* `ScriptModule` for modules with script code.
* `NativeModule` for current code in packages loaded into the global namespace - replace with your plugin later on, essentially loading the plugin under the global namespace
* `PluginModule` for your plugins in native code, loaded into modules namespace

This will unify everything very nicely, and a single `call_fn` method on the trait.  When I do a function call, I just scan through the list of modules one by one, calling `call_fn` on each, until one returns a result.  This way, I no longer need to care about where the function comes from.

jhwgh1968 commented:
> I always have difficulty understanding the 'lazy-loading" bit and why it prevents things from being pre-generated and pre-compiled.

I am sorry for imprecision in my language. It is difficult to know what to focus on with my naming and word choice, given our sprawling and meandering conversation.

The laziness was not for _loading_, but for _looking up names_. That was not a strict requirement, so much as a "I can't figure out how to deal with all this hashing stuff and heap allocated stuff" requirement for _me_ to get it working.

To try and refocus a bit, let me go back to the original test plugin in my PR:

```
// A simple custom plugin. This should not usually be done with hand-written code.
#[derive(Copy, Clone)]
struct AddOffsetPlugin(INT);
    
impl AddOffsetPlugin {
    fn add_offset(&self, x: INT) -> Dynamic {
        Dynamic::from(x + self.0)
    }
}
    
impl Plugin for AddOffsetPlugin {
    fn register_contents(self, engine: &mut Engine) -> (String, Box<dyn LazyModuleDispatcher>) {
        engine.register_fn("get_mystic_number", || { 42 as INT });
        (format!("Add Offset::{}", self.0), Box::new(self))
    }
}
    
impl LazyModuleDispatcher for AddOffsetPlugin {
    fn call<'e>(&'e self,
            fn_name: &str,
            mut args: Box<dyn Iterator<Item=Dynamic> + 'e>
    ) -> Option<Result<Dynamic, EvalAltResult>> {
        if fn_name == "add_offset" {
            if let Some(x) = args.next().and_then(|b| b.try_cast::<INT>()) {
                Some(Ok(self.add_offset(x)))
            } else {
                None
            }
        } else {
            None
        }
    }   
    fn clone_boxed(&self) -> Box<dyn LazyModuleDispatcher> {
        Box::new(*self)
    }
}
```

This is the structure I want to achieve:
* A trait in Rhai to define a plugin
* User code as regular Rust functions, and later perhaps types
* Procedural macro code to generate `AddOffsetPlugin`, a type which implements the trait and pastes in calls to the user code.

So:

> I think I consistently fail to understand why handles cannot simply be pre-generated (or generated once during module loading time) for all functions in the module. You just need to avoid having your proc macro generate a large match statement on function names and argument types, but instead generate code to hash the function, then insert it into a HashMap.

The answer is purely: I couldn't figure it out. If you can re-write it so that:

1. A HashMap is kept in the plugin struct
2. The `register_contents` will hash everything before returning itself
3. The `call` impl will look at the hash map instead of a switch

I would be fine with that, if the test passes. I just can't do it.

jhwgh1968 commented:
The sprawl and volume of our conversations is also why I am having trouble with the other thread. I feel like we are getting further away from what I have in mind, rather than closer.

I'll probably stick to this one for a while.

schungx commented:
> The laziness was not for _loading_, but for _looking up names_.

This is the exact place where I fail to understand.  Looking up a function by name is always necessary, because Rhai is a dynamic language and not statically linked.  In this definition, all function calls in Rhai are lazy.  The hashing is simply a means to speed up the lookup process by avoiding massive string comparisons (because you'll need to compare the parameter types also).

> if fn_name == "add_offset" {

This is exactly the lookup that hashing is trying to prevent.  Essentially what you're doing is, for every single function call, scan through the _entire_ plugins space (all those that are loaded).  Then match strings one by one until you find one that matches.  Repeat.

I'll propose a slight modification to your scheme below.

schungx commented:
This is a slightly modified version of `LayModuleDispatcher`:

```rust
// A simple custom plugin. This should not usually be done with hand-written code.
#[derive(Copy, Clone)]
struct AddOffsetPlugin(INT);
    
impl AddOffsetPlugin {
    fn add_offset(&self, x: INT) -> Dynamic {
        Dynamic::from(x + self.0)
    }
}
    
impl Plugin for AddOffsetPlugin {
    fn register_contents(self, engine: &mut Engine) -> (String, Box<dyn LazyModuleDispatcher>) {
        engine.register_fn("get_mystic_number", || { 42 as INT }); // plugins can have side effects too
        (format!("Add Offset::{}", self.0), Box::new(self))
    }
}

use crate::utils::some_magic_fn_provided_by_rhai;

impl LazyModuleDispatcher for AddOffsetPlugin {
    fn make_key(&self) -> u64 {
        // Use the function name and parameter types list to generate a key
        some_magic_fn_provided_by_rhai("add_offset", &[TypeId::of::<INT>()])
    }

    // Just return result.  When it gets to this point, the function call is already correct.
    fn call<'e>(&'e self,
            fn_name: &str,
            mut args: Box<dyn Iterator<Item=Dynamic> + 'e>
    ) -> Result<Dynamic, Box<EvalAltResult>> {    // Box the error to make return footprint small
            if let Some(x) = args.next().and_then(|b| b.try_cast::<INT>()) {
                Ok(self.add_offset(x))
            } else {
                Err(...)
            }
    }   
    fn clone_boxed(&self) -> Box<dyn LazyModuleDispatcher> {
        Box::new(*self)
    }
}
```

jhwgh1968 commented:
Thanks for the clarification, and the re-design! That actually helps a lot! And, the resulting form looks like something I know how to generate procedural macro code for. Exporting an entire Rust type might be trickier, but I'll deal with that when I get there.

I will happily re-write my PR to look like that, renaming `LazyModuleDispatcher` to `PluginFunction`. I should be able to do some merging as you suggest afterwards.

By the way, the `plugins` branch seems to have a different set of changes on it than I was expecting (e.g. in the other pull request to master). Could you make sure it's up to date, so I can build my PR on top of it?

schungx commented:
> By the way, the `plugins` branch seems to have a different set of changes on it than I was expecting (e.g. in the other pull request to master). Could you make sure it's up to date, so I can build my PR on top of it?

OK!

jhwgh1968 commented:
I've tried to do a little merging, but did not get too far. Partly due to implementation, partly due to uncertainty about direction. I have updated my PR with a non-functional WIP.

This version has a unit test which illustrates two different approaches for a Plugin to register a function.

1. For `euclidian_distance`, I wrapped the native function in an `Rc<Any>` closure which does the type conversions. While this would probably work within your current framework and is a small change, I find it less ergonomic and more repetitive. In particular, despite declaring all the `TypeId`s of the arguments, the closure has to cast them all anyway.

2. A native function with the new `PluginFunction` trait. I added a new method `is_varadic` just to demonstrate what power I think this could have to really expand the language. You can't write a varadic function in Rhai, but with an engine tweak or two, you possibly could.

Also note: I've had to hack the visibility of a number of items in order to get this working. I think that will need to be rearranged in order to allow plugins to be written outside of Rhai itself.

What do you think? I think option 2 will be lower overhead in the end if it could be added to modules (that's the part I'm stuck on), but I will give up on it and go with 1 if you don't have an easy fix.

schungx commented:
Let me read thru your examples first.  If at all possible, let's do No.2 which I believe will be faster in terms of implementation.

My idea is to eventually replace all modules/packages code with your plugins system, thus everything is a plugin.  Right now, packages are already implemented as modules, and the modules code is isolated so that I can replace it with an interface to your plugin later.

schungx commented:
I think it is quite fine, actually!  Some catches:

Firstly, a function accepting variant types.  This is going to be against the entire Rhai architecture, otherwise Rhai can support generics without having to manually register implementations.  The core of Rhai's function-calling mechanism is detecting the _actual_ types of parameters (wrapped in `Dynamic`) and call the correct function.

Secondly, variable number of arguments.  This can probably be done easier, as I can simply do some magic on the parser side to convert an arguments list into a `&[ Dynamic ]`.  Functions are already called this way, except that right now the individual arguments are extracted and passed by position.  There needs to be a setting marking a function as variadic, and simply omit this unwrapping.

So, if we make variadic functions always taking `&[ Dynamic ]`, this will solve your problem.

However, in your example, I still think it'll be faster if the user registers multiple copies of the same generic function with supported types (in your example you supports `INT` and `FLOAT`).  Instead of dynamically switching between the two inside the same function, which is going to be slow, register two versions of a single generic function.

jhwgh1968 commented:
Hmm. I was thinking of particular int/float promotion behavior when I wrote it, which is why I did it that way... but really, varadic functions are actually a secondary concern. I was really hoping to demnostrate that plugins could do things that Rhai itself did not support (with minimal core improvement).

Based on your comments, I will put `sum` aside for the moment, re-write `euclidean_distance` into Option 2 format, and finish up the module and engine work needed to make that work.

The next version will be a bit messy, but should be merge ready (or at least pretty close). In particular, I hope once it merges, it will be enough for you to start migrating internal things to use plugins.

schungx commented:
> I was thinking of particular int/float promotion behavior when I wrote it

Ah.  Now I understand.  That's a good thought.

Right now, one of the pain points of Rhai is supporting only one integer type and one floating point type, with no automatic conversions between the two, and no automatic upcasts for other integer types.

This is not at all uncommon in scripting languages.  For most modern scripting languages, there is only _one_ number type - floating point.  Rhai is unusual in that it supports two...

On the other hand, bringing automatic up-conversion for number types will require an expansion of the manual (i.e. the `README`) and reduce speed of most numeric operations because the engine always needs to do type checks every time.

jhwgh1968 commented:
I have updated my WIP. Now my problem is, the hashing does not cause my function to be found.

When I put in debug statements for the hashing and the lookups, it seems that the hashes never appeared anywhere the engine was looking.

This is admittedly a pretty simple implementation, but the hashing infrastructure made me think I didn't need anything else. Do you have any idea what I missed?

<details>

    <snip>
    
    [src/module.rs:274] (&name, hash_fn) = (
        "trim",
        13367664171665135343,
    )
    [src/module.rs:274] (&name, hash_fn) = (
        "get_mystic_number",
        2080782379631061082,
    )
    [src/module.rs:274] (&name, hash_fn) = (
        "euclidean_distance",
        5861223951096452991,
    )
    [src/module.rs:529] (name, hash_fn_native) = (
        "euclidian_distance",
        12541146500469585741,
    )
    [src/module.rs:530] self.all_functions.keys().collect::<Vec<_>>() = [
        10908691540684989219,
    ]
    [src/module.rs:529] (name, hash_fn_native) = (
        "euclidian_distance",
        4519245633418098358,
    )
    [src/module.rs:530] self.all_functions.keys().collect::<Vec<_>>() = [
        10908691540684989219,
    ]
    Error: ErrorFunctionNotFound("euclidian_distance", 0:0)

</details>

schungx commented:
There is one mega hashing function instead of different ones for different occasions, so you'll have to pass in the correct parameters.

```
calc_hash_fn( None, fn_name, num_args, Iterator<TypeId>)
```

This is what you'll want to pass in for a straight function with known number and types of arguments.

schungx commented on src/plugin.rs:
```diff
@@ -0,0 +1,35 @@
+    fn is_method_call(&self) -> bool;
+    fn is_varadic(&self) -> bool;
+
+    fn call(&self, args: &[&mut Dynamic], pos: Position) -> Result<Dynamic, EvalAltResult>;
```
Better return `Box<EvalAltResult>` to be consistent.

schungx commented on src/result.rs:
```diff
@@ -75,6 +75,8 @@ pub enum EvalAltResult {
     /// Arithmetic error encountered. Wrapped value is the error message.
     ErrorArithmetic(String, Position),
+    /// Parameter was incorrectly typed.
+    ErrorFnInputType(String, Position),
```
I wonder where this is going to hit... parameters are never supposed to be of the incorrect type...

jhwgh1968 replied:
That was a left-over from my varadic stuff. I'll remove it.

jhwgh1968 commented:
> There is one mega hashing function instead of different ones for different occasions, so you'll have to pass in the correct parameters.

I just called `set_fn`; isn't that the correct parameters?

```
pub fn set_fn(&mut self, name: String, access: FnAccess, params: &[TypeId], func: CF) -> u64 {
    let hash_fn = calc_fn_hash(empty(), &name, params.len(), params.iter().cloned());
[...]
    hash_fn
}
```

schungx commented:
Yesh, the code looks right to me, except that it is not finding your function.  Let me take a closer look.

jhwgh1968 commented:
Both of your other changes were fixed.

schungx commented:
Can you try `euclidian_distance` without the `get_mystic_number()` call?

schungx commented:
`m.set_fn("euclidean_distance".to_string(), FnAccess::Public,`

`math::euclidian_distance`

Are you sure you didn't make a typo?

jhwgh1968 commented:
> Are you sure you didn't make a typo?

Where is that Captain Picard facepalm gif...

Thanks. Now it panics, and I know why. I'll keep working on it.

jhwgh1968 commented:
I just pushed a version where the unit test passes! :tada: 

Once I clean up some types, and make it work with all the other feature flag combinations, it will be ready for final review.

schungx commented:
> But, based on your logs, it seems that get_qualified_fn is called twice. I'll need to find out why.

OK, that's normal.  Once to check scripted functions, once to check native functions.

schungx commented:
I'll catch up `plugins` with latest.

jhwgh1968 commented:
I have updated my PR.

Instead of my own feature gate, I put everything behind the `no_module` feature gate to avoid interactions. It is now also compatible with the `sync feature gate, which was not true before.

It is ready to merge, after (I hope) one final review.

schungx commented:
Looks good.  I would suggest removing the `no_module` gates because when we re-implement the basic library as plugins we'd need it anyway.  Just avoid exposing public plugin API's is enough.

jhwgh1968 commented:
Done!

schungx commented:
Just some minor suggestions and I think this can be merged in.

However, I am thinking of keeping the `no_module` feature to disable loading custom modules only.  That means removing most `no_module` feature gates as in your PR, but keeping it to remove the public module API's (e.g. loading modules into `scope`).

I'll start doing this removal on the `master` branch to keep the `plugins` branch in sync.

schungx commented:
@jhwgh1968   can you rebase this PR on the latest?  This will ensure a smooth merge.  I've also removed many `no_module` gates in `master` to prepare for eventual merging `plugins` back to `master`.

However, pls keep the `no_module` feature for the time being.  In the latest, it only affects public-facing API's.

schungx commented:
@jhwgh1968   can you rebase this PR on the latest?  This will ensure a smooth merge.  I've also removed many `no_module` gates in `master` to prepare for eventual merging `plugins` back to `master`.

However, pls keep the `no_module` feature for the time being.  In the latest, it only affects public-facing API's.

jhwgh1968 commented:
Sorry, I took your "remove it" a little too literally! I have rebased and fixed that, so it should be good now.

schungx commented:
A couple of comments:

1) I have a few comments on removing multiple versions of code based on the `sync` gate.

2) For most of the remaining `no_module` gates you need to keep them there (i.e. undelete them).  In some places they're necessary otherwise it won't compile with `no_module`.  In other places they restrict the public module API's.

I'll merge this in once you take care of them.

jhwgh1968 commented:
> For most of the remaining no_module gates you need to keep them there (i.e. undelete them).

Done.

> I have a few comments on removing multiple versions of code based on the sync gate.

Where are they? I didn't see a full review be opened. I'd be happy to fix them.

schungx commented:
> Where are they? I didn't see a full review be opened. I'd be happy to fix them.

Well...  I just commented them directly into the changed files...  Not sure about this code review thing...

schungx commented on src/any.rs:
```diff
@@ -5,7 +5,7 @@ use crate::parser::{ImmutableString, INT};
 use crate::r#unsafe::{unsafe_cast_box, unsafe_try_cast};
 
 #[cfg(not(feature = "no_float"))]
-use crate::parser::FLOAT;
```
Better keep `crate::parser::FLOAT` because that frees us in the future to remap anything under the `crate::` root.

jhwgh1968 replied:
I changed this because it gave me a compiler error before. Now it doesn't, so I will fix it (and the two other instances as well).

schungx commented on src/fn_native.rs:
```diff
@@ -131,6 +154,21 @@ impl CallableFunction {
+    /// Panics if the `CallableFunction` is not `Plugin`.
+    pub fn get_plugin_fn<'s>(&'s self) -> SharedPluginFunction {
+        match self {
+            #[cfg(not(feature = "sync"))]
```
`Self::Plugin(f) => f.clone()` avoids the multiple versions.

schungx commented on src/fn_native.rs:
```diff
@@ -141,4 +179,16 @@ impl CallableFunction {
+
+    #[cfg(feature = "sync")]
+    /// Create a new `CallableFunction::Plugin`.
+    pub fn from_plugin(plugin: impl PluginFunction + 'static + Send + Sync) -> Self {
```
Use `Self::Plugin(plugin.into())` to avoid multiple versions.

jhwgh1968 replied:
Not quite:
```
182 |         Self::Plugin(plugin.into())
    |                      ^^^^^^^^^^^^^ the trait `std::convert::From<impl PluginFunction + 'static + Send + Sync>` is not implemented for `std::rc::Rc<dyn plugin::PluginFunction>`
    |
    = note: required because of the requirements on the impl of `std::convert::Into<std::rc::Rc<dyn plugin::PluginFunction>>` for `impl PluginFunction + 'static + Send + Sync`
```

schungx replied:
Ah!  OK!

schungx commented on src/fn_register.rs:
```diff
@@ -156,10 +183,9 @@ pub fn by_value<T: Variant + Clone>(data: &mut Dynamic) -> T {
 
-#[cfg(feature = "plugins")]
-impl<PL: Plugin> RegisterPlugin<PL> for Engine {
+impl<PL: crate::plugin::Plugin> RegisterPlugin<PL> for Engine {
```
For consistent styling, let's import all the namespaces at the top of the file.  `import crate::plugin::Plugin;`

schungx commented on src/lib.rs:
```diff
@@ -75,10 +75,14 @@ mod engine;
 mod fn_call;
 mod fn_func;
-mod fn_native;
+pub mod fn_native;
```
Avoid publicly exposing entire sub-modules.  Better use `pub use` later on to selectively export.

jhwgh1968 replied:
I did it this way, because (a) almost everything was `pub`, and (b) everything that was `pub` was needed for actual plugin writing.

It seems much easier for module authors to import `rhai::fn_native::*` than the number of things that would need to be imported, including several traits.

jhwgh1968 replied:
I wouldn't mind adding this to the `plugin` module as public exports either, so that they could import `rhai::plugin::*` to get everything. But the general idea is the same.

schungx replied:
Hhmmm... if you think you'll eventually need all public stuff under `plugins` then it makes sense.

However, you'll need to be careful to `pub(crate)` the stuff you don't want to expose...

schungx commented on src/lib.rs:
```diff
@@ -75,10 +75,14 @@ mod engine;
 mod fn_register;
 mod module;
 mod optimize;
+#[cfg(not(feature = "no_module"))]
```
Not sure if you'd need to expose the entire `plugin` module.  Better to selectively use `pub use` to expose certain subset.

jhwgh1968 replied:
The comment above on the other `pub mod` also applies here.

jhwgh1968 commented:
All resolved comments are now fixed, the rest replied to.

jhwgh1968 commented:
Thanks for the merge, @schungx! I'll open a new PR for the next step, and clean up the `plugin` export a bit as well.

schungx commented:
Thanks!

schungx commented:
I'll sync it up with `master`.
